tools/build_script_generator/cmake/resources/CompilerWarnings.cmake.in

```diff
@@ -0,0 +1,58 @@
+# from here:
+#
+# https://github.com/lefticus/cppbestpractices/blob/master/02-Use_the_Tools_Available.md
+
+function(set_project_warnings project_name)
+  option(WARNINGS_AS_ERRORS "Treat compiler warnings as errors" TRUE)
+
+  set(CLANG_WARNINGS
+      -Wall
+      -Wextra # reasonable and standard
+      -Wshadow # warn the user if a variable declaration shadows one from a parent context
+      -Wnon-virtual-dtor # warn the user if a class with virtual functions has a non-virtual destructor. This helps
+                         # catch hard to track down memory errors
+      -Wold-style-cast # warn for c-style casts
```

Comment:
All these flags must come from the build system generator (to be identical for all build systems)

Reply:
Yes it can do both. You can specify _modm_warnings_ or _projects_warnings_. Is that a problem?

Reply:
The actual warnings will be mildly different depending on target (eg. `-Wdouble-promotion` depends on FPU type), so this file needs to be generated too.

Reply:
CompilerWarnings.cmake applies to user code only if used at all. lbuild is not overwriting this file if it exists. I see a usefulness if someone wants to stick to best practice and some warnings is broken in modm like -Wpedantic.  Do we keep it?

Reply:
I think this idea is great! It just needs to be made available in a sensible way to the user in both CMake and SCons.

I would have to think about which warnings are necessary for the *whole* code and which are merely "cosmetic" warnings for modm. For example, `-Wdouble-promotion` is important for the whole application code to not accidentally pull in software math routines for double-width floats, otherwise you loose the benefits of the FPU. But `-Wpendantic` should definitely be optional.

It would also make sense to give different warnings to code like CMSIS-DSP or TinyUSB, which are currently throwing a lot of warnings. We already use different [per-file compile flags for CMSIS-DSP](https://github.com/modm-io/modm/blob/develop/ext/arm/dsp.lb#L48-L85) so the build system generator supports per-file flags already.